crates/turborepo-lib/src/daemon/connector.rs

```diff
@@ -0,0 +1,548 @@
+        path::{Path, PathBuf},
+    };
+
+    use sysinfo::Pid;
+
+    use super::*;
+
+    #[cfg(not(target_os = "windows"))]
+    const NODE_EXE: &str = "node";
+    #[cfg(target_os = "windows")]
+    const NODE_EXE: &str = "node.exe";
+
+    fn pid_path(tmp_path: &Path) -> PathBuf {
+        tmp_path.join("turbod.pid")
+    }
+
+    fn sock_path(tmp_path: &Path) -> PathBuf {
+        tmp_path.join("turbod.sock")
+    }
+
+    #[tokio::test]
+    async fn handles_invalid_pid() {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let tmp_path = tmp_dir.path().to_owned();
+
+        let pid = pid_path(&tmp_path);
+        std::fs::write(&pid, "not a pid").unwrap();
+
+        let connector = DaemonConnector {
+            pid_file: pid,
+            sock_file: sock_path(&tmp_path),
+            can_kill_server: false,
+            can_start_server: false,
+        };
+
+        assert_matches!(
+            connector.get_or_start_daemon().await,
+            Err(DaemonConnectorError::NotRunning)
+        );
+    }
+
+    #[tokio::test]
+    async fn handles_missing_server_connect() {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let tmp_path = tmp_dir.path().to_owned();
+
+        let pid = pid_path(&tmp_path);
+        let sock = sock_path(&tmp_path);
+
+        let connector = DaemonConnector {
+            pid_file: pid,
+            sock_file: sock,
+            can_kill_server: false,
+            can_start_server: false,
+        };
+
+        assert_matches!(
+            connector.connect().await,
+            Err(DaemonConnectorError::NotRunning)
+        );
+    }
+
+    #[tokio::test]
+    async fn handles_kill_dead_server_missing_pid() {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let tmp_path = tmp_dir.path().to_owned();
+
+        let pid = pid_path(&tmp_path);
+        let sock = sock_path(&tmp_path);
+
+        let connector = DaemonConnector {
+            pid_file: pid,
+            sock_file: sock,
+            can_kill_server: false,
+            can_start_server: false,
+        };
+
+        assert_matches!(
+            connector.kill_dead_server(Pid::from(usize::MAX)).await,
+            Ok(())
+        );
+    }
+
+    #[tokio::test]
+    async fn handles_kill_dead_server_missing_process() {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let tmp_path = tmp_dir.path().to_owned();
+
+        let pid = pid_path(&tmp_path);
+        std::fs::write(&pid, usize::MAX.to_string()).unwrap();
+        let sock = sock_path(&tmp_path);
+        std::fs::write(&sock, "").unwrap();
+
+        let connector = DaemonConnector {
+            pid_file: pid,
+            sock_file: sock,
+            can_kill_server: false,
+            can_start_server: false,
+        };
+
+        assert_matches!(
+            connector.kill_dead_server(Pid::from(usize::MAX)).await,
+            Ok(())
+        );
+
+        assert!(!connector.pid_file.exists(), "pid file should be deleted");
```

Comment:
When the pid lock is opened, the library deletes the lock if there is no process that owns it. Should we change that behaviour?

Reply:
As discussed, yes, let's avoid deleting the pid file from the client to avoid race conditions. In the future we can add a flag or subcommand to allow the user to specify that we should attempt a cleanup of bad or stale files.